File: Data Structures & Algorithms/max-water-container/submission-1.java
class Solution {
  public int maxArea(int[] heights) {
    int left = 0;
    int right = heights.length - 1;

    int most = 0;
    int area = 0;
    while (left != right) {
      area = Math.min(heights[left], heights[right]) * (right - left);

      most = Math.max(area, most);
      
      if (heights[left] < heights[right]) 
        left++;

      else 
        right--;
    }

    return most;
  }
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-1.cpp
class Solution {
public:
    bool isBalanced(TreeNode* root) {
      bool balanced = true;
      std::unordered_map<TreeNode*, int> map;
      std::stack<TreeNode*> stack;
      map.insert({nullptr, 0});
      stack.push(root);


      while (!stack.empty()) {
        TreeNode* current = stack.top();
        if (current == nullptr) {stack.pop(); continue;}
        if (map.find(current->left) == map.end()) {
          stack.push(current->left);
        } else if (map.find(current->right) == map.end()) {
          stack.push(current->right);
        } else {
          int leftHeight = map.at(current->left);
          int rightHeight = map.at(current->right);

          if (std::abs(rightHeight - leftHeight) > 1) balanced = false;

          map.insert({current, std::max(rightHeight, leftHeight) + 1});
          stack.pop();
        }
      }

      return balanced;
    }
};
